### PR TITLE
259 homepage infinite scrolling v2

### LINKED ISSUE #259: [259] Infinite scrolling on homepage
**Current state**
When the user scrolls to the end of the posts in the homepage, scrolling will not be possible.

**Expected state**
When the user scrolls to the end of the posts in the homepage, older posts get displayed again so the user can scroll through content infinitely.

**TO DO**
- [x] Fetch and show older posts when end is reached

**Necessities**
#216 

**Branch**
259-homepage-infinite-scrolling

### PR DESCRIPTION
**Pull request by jerohero**
closes #259 

<br />

 **What is it supposed to do**  
Fixes the issue from the previous infinite scrolling PR ( #275 ) where the previous post blocks wouldn't fill when loading more posts.

Current state: 
image

New state:
image

<br />

 **How can we test your branch**  
Try adding/removing posts and scroll down (home page) to see if post blocks get filled properly, and if there is no unwanted behavior like posts disappearing.

<br />

 **Additional notes**  
Since posts get fetched in pages of 6 posts, the pages won't end 'cleanly' on a fetch when one page doesn't give 6 posts (they will only get filled at the next fetch, when the user scrolls down):
image
If this is undesired I could fix this by simply fetching again when the fetch results don't contain 6 posts, so the user doesn't get to see it as much. I could also increase the page length to get the same result (which will most likely be done in the future anyway as fetching once every 6 posts can be expensive). Note that this only occurs once the end of the homepage is reached, which I assume won't happen often in production.

<br />

**1. General checklist**
- [x] a. Tested happy flow.
- [x] b. Tested unhappy flow.
- [x] c. No unexpected exceptions.
- [x] d. No code with high complexity. (Big O)
- [x] e. EOF newline.
- [x] f. No linter warnings or errors.

<br />

**2. If relevant, front-end checklist**
- [x] a. Responsive. (Desktop all the way to mobile 320px)
- [x] b. Using SCSS variables where relevant.
- [x] c. Relevant class names
- [x] d. No modified global SCSS properties, except if necessary and fully tested the impact.
- [x] e. Checked [caniuse](https://caniuse.com) for more modern CSS properties.

<br />

**3. If relevant, back-end checklist**
- [ ] a. No unhandled promise rejections.
- [ ] b. Seeders still functional.
- [ ] c. No primary-key violations.
- [ ] d. Normalization, at least [3NF](https://en.wikipedia.org/wiki/Database_normalization).
- [ ] e. Used guards (middleware) where necessary and tested if route is guarded.

<br />

**4. If relevant, test these browsers**
- [x] a. Chrome
- [x] b. Firefox
- [ ] c. Opera
- [x] d. Edge
- [ ] e. Safari

### CODE REVIEW

File: frontend/src/components/posts/index.js [hunk 1]
```diff
@@ -75,10 +71,21 @@ class Posts extends Component {
...
-    if (this.state.endReached) {
-      return
-    }
+	  if (this.state.endReached) {
+		  this.setCurrentPage(0)
+		  this.setEndReached(false)
+		  this.fetchPosts()
+
+		  return
+	  }
```
Comment (AjobK) [inline frontend/src/components/posts/index.js]: Wrong indentation. Two spaces

File: frontend/src/components/posts/index.js [hunk 2]
```diff
@@ -95,16 +102,61 @@ class Posts extends Component {
...
+
+    while (true) {
+      if (currentElement.props.children) {
+        currentElement = currentElement.props.children
+      } else {
+        return currentElement.props.posts
+          ? currentElement.props.posts
+          : []
+      }
+    }
```
Comment (AjobK) [inline frontend/src/components/posts/index.js]:
```suggestion
    while (currentElement?.props?.children) {
        const { children } = currentElement.props
        currentElement = children || []
    }
    
    return currentElement
```

File: frontend/src/components/posts/index.js [hunk 3]
```diff
@@ -95,16 +102,61 @@ class Posts extends Component {
...
+
+    if (remainingPosts.length > 0) {
+      this.renderNewPostsBlock(remainingPosts)
+    }
+  }
+
+  renderNewPostsBlock(posts) {
     let singleLi = document.createElement('li')
 
     singleLi.classList.add(styles.post)
```
Comment (AjobK) [inline frontend/src/components/posts/index.js]: Not using this?

File: frontend/src/components/posts/index.js [hunk 4]
```diff
@@ -95,16 +102,61 @@ class Posts extends Component {
...
 
-    let postsBlocks = this.state.postsBlocks
+    const postsBlocks = this.state.postsBlocks
```
Comment (AjobK) [inline frontend/src/components/posts/index.js]:
```suggestion
    const { postsBlocks } = this.state
```

File: frontend/src/components/posts/index.js [hunk 5]
```diff
@@ -39,21 +39,17 @@ class Posts extends Component {
...
-
-        this.setState({
-          page: this.state.page + 1,
-        })
+        this.setCurrentPage(this.state.page + 1)
         this.renderNewPosts(json.posts ? json.posts : [])
 
         if (json.posts.length < this.MAX_POSTS_IN_BLOCK) {
           this.setEndReached(true)
         }
```
Comment (AjobK) [inline frontend/src/components/posts/index.js]: Make more dynamic. This would fail if back-end pagination were to set larger portions per page.

**Review by AjobK — CHANGES_REQUESTED** [review]
image

Pagination doesn't seem to work well when changing on back-end to 13 per page.

Check rest of comments too.

File: frontend/src/components/posts/index.js [hunk 6]
```diff
@@ -39,21 +39,17 @@ class Posts extends Component {
...
-          this.setEndReached(true)
-
-          return
-        }
-
-        this.setState({
-          page: this.state.page + 1,
-        })
+        this.setCurrentPage(this.state.page + 1)
         this.renderNewPosts(json.posts ? json.posts : [])
```
Comment (jerohero) [inline frontend/src/components/posts/index.js]: json.posts? || []

Comment (jerohero) [reply frontend/src/components/posts/index.js]: image
No worky

Comment (jerohero) [reply frontend/src/components/posts/index.js]: json.posts || [] does

File: frontend/src/components/posts/index.js (hunk 2 above)
Comment (jerohero) [reply frontend/src/components/posts/index.js]: _Return currentElement_ had to be _return currentElement.props.posts_. Unfortunately had to spend 3 hours to see this 😅

**Comment by jerohero** [comment]
> image
> 
> Pagination doesn't seem to work well when changing on back-end to 13 per page.
> 
> Check rest of comments too.

Fixed it with the following code in _renderNewPostsAtLastPostsBlock():_
```
const amountOfPostBlocks = Math.ceil(posts.length / this.MAX_POSTS_IN_BLOCK)

for (let i = 0; i < amountOfPostBlocks; i++) {
  const postsInPostsBlock = posts.slice(
    this.MAX_POSTS_IN_BLOCK * i,
    this.MAX_POSTS_IN_BLOCK * (i + 1)
  )

  postsBlocks.push(this.createPostsBlock(postsInPostsBlock))
}
```

_MAX_POSTS_IN_BLOCK_ will still always be 6 since that's how they're shown in the post blocks, but the logic behind deciding whether the end of the posts has been reached (in _fetchPosts()_) has been changed to:
```
if (json.posts.length < json.per_page) {
  this.setEndReached(true)
}
```

This makes it dynamic; the posts amount per page from the API is used.

One thing I did notice while testing this is that it doesn't work when the posts per page amount is 1. I don't think we'll ever want to set it to that amount, but it might be worth mentioning anyway.

**Review by AjobK — CHANGES_REQUESTED** [review]
As discussed in call, make it work with refactored pagination

**Comment by jerohero** [comment]
image
Tests succeed

**Comment by jerohero** [comment]
@AjobK I made the changes we discussed last week (fix thumbnails and scroll end reached detection). I fixed an issue that suddenly made scrolling not function at all, so I am not entirely sure this is what you meant.

**Comment by MrGooss** [comment]
Issue with page infinitely reloading when there aren't enough posts to fill the page. Could be fixed by checking the total count of posts en seeing if loading new posts is necessary or not.

File: frontend/src/components/posts/index.js [hunk 7]
```diff
@@ -95,16 +101,63 @@ class Posts extends Component {
...
-    postsBlocks.push(this.createPostsBlock(posts))
+    if (!postsBlock)
+      return []
+
+    while (currentElement?.props?.children) {
+      const { children } = currentElement.props
+      currentElement = children || []
+    }
+
+    return currentElement.props.posts
```
Comment (MrGooss) [inline frontend/src/components/posts/index.js]: Might cause an issue with props/posts being undefined when it gets set to []

**Review by MrGooss — CHANGES_REQUESTED** [review]

**Review by AjobK — APPROVED** [review]

**Pull request merged**